photo-effects/src/layout/Canvas/components/CanvasArea/CanvasItem.js

```diff
@@ -1,135 +1,225 @@
 import React, { Component } from 'react';
 import ReactDOM from 'react-dom';
 
-import Resizer from './Resizer';
-
 export default class CanvasItem extends Component {
   state = {
-    drag: false,
-    resize: false,
+    resizing: false,
+    x: 0,
+    y: 0,
     max_x: 0,
     max_y: 0,
-    width: 0,
-    height: 0,
-    resizeWidth: null,
-    resizeHeight: null
+    item: {}
   }
 
-  componentDidMount() {
+  componentDidMount() { 
     const _this = ReactDOM.findDOMNode(this);
     const node = _this.getBoundingClientRect();
+    const { left, top } = node;
     const parent = _this.offsetParent.getBoundingClientRect();
-    
+    const { width, height } = parent;
+    const { nodeWidth, nodeHeight, item, z } = this.props; 
+
     this.setState({ 
-      max_x: parent.width - node.width,
-      max_y: parent.height - node.height,
-      width: node.width,
-      height: node.height
+      max_x: width - nodeWidth,
+      max_y: height - nodeHeight,
+      item: {
+        ...item, left, top
+      }
     })
   }
 
-  item = () => {
-    let top = 0;
-    if(this.state.drag) {
-      if(this.props.new_y - (this.state.height / 2) > this.state.max_y) {
-        top = this.state.max_y
-      } else if(this.props.new_y - (this.state.height / 2) < 0) {
-        top = 0
-      } else {
-        top = this.props.new_y - (this.state.height / 2)
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if(!nextProps.resizing) {
+      return {
+        item: {
+          width: nextProps.nodeWidth,
+          height: nextProps.nodeHeight
+        }
       }
-    } else if(this.props.y - (this.state.height / 2) <= 0) {
-      top = 0
-    } else {
-      top = this.props.y - (this.state.height / 2)
-    }
+    } else return null;
+  }
+  
+
+  item = () => {
+    const { 
+      dragging, 
+      resizing, 
+      new_x, 
+      new_y,
+      max_x,
+      max_y, 
+      nodeX,
+      nodeY,
+      nodeWidth, 
+      nodeHeight,
+      hover
+    } = this.props;
+
+    const {
+      item
+    } = this.state;
+
+    const {
+      width,
+      height
+    } = item;
 
     let left = 0;
-    if(this.state.drag) {
-      if(this.props.new_x - (this.state.width / 2) > this.state.max_x) {
-        left = this.state.max_x
-      } else if(this.props.new_x - (this.state.width / 2) < 0) {
-        left = 0;
-      } else {
-        left = this.props.new_x - (this.state.width / 2)
-      }
-    } else if(this.props.x - (this.state.width / 2) <= 0) {
-      left = 0
+    let top = 0;
+
+    if(dragging) {
```

Comment:
Hey Doug can you define what dragging is here? Or create functions that will better help others to maybe understand what is happening? Just a suggestion!